src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,132 @@
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel : INotifyPropertyChanged, IDisposable
+    {
+        private readonly ILanguageServiceBroker2 _languageServiceBroker;
+        private readonly VisualStudioCodeWindowInfoService _visualStudioCodeWindowInfoService;
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        public DocumentOutlineViewModel(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener,
+            VisualStudioCodeWindowInfoService visualStudioCodeWindowInfoService,
+            CompilationAvailableTaggerEventSource textViewEventSource)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _textViewEventSource = textViewEventSource;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
+            _sortOption = SortOption.Location;
+            _documentSymbolUIItems = new ObservableCollection<DocumentSymbolUIItem>();
+
+            // setup work queues
+            _documentSymbolQueue = new AsyncBatchingWorkQueue<VisualStudioCodeWindowInfo, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                EqualityComparer<VisualStudioCodeWindowInfo>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _filterAndSortQueue = new AsyncBatchingWorkQueue<FilterAndSortOptions, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<FilterAndSortOptions>.Default,
+                asyncListener,
+                CancellationToken);
+
+            _updateUIQueue = new AsyncBatchingWorkQueue<UIData>(
+                DelayTimeSpan.NearImmediate,
+                UpdateUIAsync,
+                asyncListener,
+                CancellationToken);
+
+            _textViewEventSource.Changed += OnEventSourceChanged;
+            _textViewEventSource.Connect();
+
+            // queue initial model update
+            EnqueueModelUpdateAsync().Forget();
+        }
+
+        private string? _searchText;
+        public string? SearchText
+        {
+            get
+            {
```

Comment:
these all need threading asserts.

Reply:
They do not, they can be read and written to from any thread. `INotifyPropertyChanged` handles UI updates.

Reply:
we can assert it can be read/written from either thread.  we have an assert precisely for that :)  it acts as a form of documentation.

Note: if this can be read/written from any thread, then we need to make these threadsafe.  right now they are racy.  two threads may come in and decide to change (Despite the other winning) or not change.